File: scala/bob/src/main/scala/Bob.scala
class Bob {

  val hey: String => String = {
    case s if isEmpty(s) => "Fine. Be that way!"
    case s if isShout(s) => "Whoa, chill out!"
    case s if isQuestion(s) => "Sure."
    case _ => "Whatever."
  }

  private def isEmpty(message: String) = message.trim.isEmpty

  private def isQuestion(message: String) = message.endsWith("?")

  private def isShout(message: String) = {
    def allUppercase = message == message.toUpperCase()
    def containsLetters = message.exists(_.isLetter)

    allUppercase && containsLetters
  }
}

File: haskell/bob/Bob.hs
{-# OPTIONS_GHC -Wall #-}

module Bob (responseFor) where

import Data.List (isSuffixOf)
import Data.Char (isUpper, isLetter, isSpace)

responseFor :: String -> String
responseFor msg
    | isBlank = "Fine. Be that way!"
    | isShout = "Whoa, chill out!"
    | isQuestion = "Sure."
    | otherwise = "Whatever."
    where
      isBlank = all isSpace msg
      isQuestion = "?" `isSuffixOf` msg
      isShout = (any isLetter msg) && (all isUpper letters)
      letters = filter isLetter msg